mongoengine/base/metaclasses.py

```diff
@@ -385,48 +385,71 @@ def __new__(mcs, name, bases, attrs):
             new_class._db_field_map[id_name] = id_db_name
             new_class._reverse_db_field_map[id_db_name] = id_name
-            # Prepend id field to _fields_ordered
+
+            # Prepend the ID field to _fields_ordered (so that it's *always*
+            # the first field).
             new_class._fields_ordered = (id_name, ) + new_class._fields_ordered
 
-        # Merge in exceptions with parent hierarchy
+        # Merge in exceptions with parent hierarchy.
         exceptions_to_merge = (DoesNotExist, MultipleObjectsReturned)
         module = attrs.get('__module__')
         for exc in exceptions_to_merge:
             name = exc.__name__
-            parents = tuple(getattr(base, name) for base in flattened_bases
-                            if hasattr(base, name)) or (exc,)
-            # Create new exception and set to new_class
+            parents = tuple(
+                getattr(base, name)
+                for base in flattened_bases
+                if hasattr(base, name)
+            ) or (exc,)
+
+            # Create a new exception and set it as an attribute on the new
+            # class.
             exception = type(name, parents, {'__module__': module})
             setattr(new_class, name, exception)
 
         return new_class
 
     @classmethod
     def get_auto_id_names(mcs, new_class):
+        """Find a name for the automatic ID field for the given new class.
+
+        Return a two-element tuple where the first item is the field name (i.e.
+        the attribute name on the object) and the second element is the DB
+        field name (i.e. the name of the key stored in MongoDB).
+
+        Defaults to ('id', '_id'), or generates a non-clashing name in the form
+        of ('auto_id_X', '_auto_id_X') if the default name is already taken.
+        """
         id_name, id_db_name = ('id', '_id')
-        if id_name not in new_class._fields and \
-                id_db_name not in (v.db_field for v in new_class._fields.values()):
+        existing_fields = new_class._fields
+        existing_db_fields = (v.db_field for v in new_class._fields.values())
+        if (
+            id_name not in existing_fields and
+            id_db_name not in existing_db_fields
+        ):
             return id_name, id_db_name
-        id_basename, id_db_basename, i = 'auto_id', '_auto_id', 0
-        while id_name in new_class._fields or \
-                id_db_name in (v.db_field for v in new_class._fields.values()):
+
+        id_basename, id_db_basename, i = ('auto_id', '_auto_id', 0)
+        while True:
             id_name = '{0}_{1}'.format(id_basename, i)
             id_db_name = '{0}_{1}'.format(id_db_basename, i)
-            i += 1
-        return id_name, id_db_name
+            if (
+                id_name not in existing_fields and
+                id_db_name not in existing_db_fields
+            ):
+                return id_name, id_db_name
+            else:
+                i += 1
```

Comment:
I wasn't aware of this `auto_id` feature but it does not looks like its working correctly
![image](https://user-images.githubusercontent.com/11854690/60047739-bc67f300-96ca-11e9-8eea-7dcc870b0820.png)

Currently no test seems to be testing the raw structure when `auto_id` is triggered

I would expect an _auto_id_0 field in the raw mongo document, am I missing something?

Reply:
Ah, I missed this :( I have checked that instantiating the document creates correct fields, i.e. in your example:
```
In [18]: MyDoc._fields
Out[18]:
{'an_int': <mongoengine.fields.IntField at 0x109e058d0>,
 'auto_id_0': <mongoengine.base.fields.ObjectIdField at 0x109e05350>,
 'id': <mongoengine.fields.IntField at 0x109e05550>}
```

However, I did *not* check persisting a document with such a field structure... Technically there's an automated test for it here: https://github.com/MongoEngine/mongoengine/blob/b1002dd4f9e5ede3514b0805cffbee29a069cda4/tests/document/inheritance.py#L388-L405

That said, it only works because `berlin.id` isn't set. **This feature is indeed broken and AFAICT, it has always been broken :(**

---

That might be a good thing after all... It means that most likely nobody is using this (overly magical and half-implemented) feature and we can gut it. I'll do that in a follow-up PR, providing an explanation of how this feature came to be, why it doesn't work correctly (hint: it all started with https://github.com/MongoEngine/mongoengine/pull/961), and why we don't really need it.

For now, I'll keep the docstring/readability improvements related to that feature, so that it's easier to understand its behavior (even if that behavior is broken).